client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -172,6 +172,52 @@ private final class PoolManager[F[_], A <: Connection[F]](
       }
     }
 
+  private def releaseRecyclable(key: RequestKey, connection: A): Unit =
+    waitQueue.dequeueFirst(_.key == key) match {
+      case Some(Waiting(_, callback)) =>
+        logger.debug(s"Fulfilling waiting connection request: $stats")
+        callback(Right(NextConnection(connection, fresh = false)))
+
+      case None if waitQueue.isEmpty =>
+        logger.debug(s"Returning idle connection to pool: $stats")
+        addToIdleQueue(connection, key)
+
+      case None =>
+        findFirstAllowedWaitor match {
+          case Some(Waiting(k, cb)) =>
+            // This is the first waiter not blocked on the request key limit.
+            // close the undesired connection and wait for another
+            connection.shutdown()
+            decrConnection(key)
+            createConnection(k, cb)
+
+          case None =>
+            // We're blocked not because of too many connections, but
+            // because of too many connections per key.
+            // We might be able to reuse this request.
+            addToIdleQueue(connection, key)
+        }
+    }
+
+  private def releaseNonRecyclable(key: RequestKey, connection: A): Unit = {
+    decrConnection(key)
+
+    if (!connection.isClosed) {
+      logger.debug(s"Connection returned was busy.  Shutting down: $stats")
+      connection.shutdown()
+    }
+
+    if (waitQueue.nonEmpty) {
+      logger.debug(s"Connection returned could not be recycled, new connection needed: $stats")
```

Comment:
We say we're opening a new connection here, but we silently do nothing if `findFirstAllowedWaiter` returns None.  I think we could replace `if (waitQueue.nonEmpty)` with `findFirstAllowedWaiter` and match a `Some` and `None`.